### PR TITLE
CTX-6592: Image segmentation YOLOv8

### PR DESCRIPTION
**Pull request by duskovesic**

### CODE REVIEW

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+from pathlib import Path
+from typing import Optional
+
+import logging
+import random
+import json
+import yaml
```
Comment (bogdant36) [inline tasks/image-segmentation-yolov8/src/dataset.py:7]: yaml is not standard python package

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+def createSampleResults(taskRun: TaskRun[ImageDataset], sampleResults: list[dict[str, str]]) -> None:
+    csvSamplesPath = folder_manager.temp / "sample_results.csv"
+
+    resultsForCsv: list[dict[str, str]] = []
+    for result in sampleResults:
+        resultsForCsv.append({index: result[index] for index in ["sample id", "sample name", "accuracy"]})
+
+    with csvSamplesPath.open("w") as file:
+        writer = csv.DictWriter(file, fieldnames = resultsForCsv[0].keys())
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:57]: Is it possible for `resultsForCsv` to be empty here?

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    else:
+        logging.info(">> [Image Segmentation] The csv file with dataset results has been uploaded as artifact")
+
+
+def createSampleResults(taskRun: TaskRun[ImageDataset], sampleResults: list[dict[str, str]]) -> None:
+    csvSamplesPath = folder_manager.temp / "sample_results.csv"
+
+    resultsForCsv: list[dict[str, str]] = []
+    for result in sampleResults:
+        resultsForCsv.append({index: result[index] for index in ["sample id", "sample name", "accuracy"]})
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:54]: I would not add this data to a list first, but rather call `writer.writerow` directly for this data

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+import numpy as np
+import torch
+
+
+def createDatasetResults(taskRun: TaskRun[ImageDataset], sampleResults: list[dict[str, str]]) -> None:
+    datasetResult: list[dict[str, str]] = []
+
+    for label in taskRun.dataset.classes.labels:
+        classAcc = [float(result[label]) for result in sampleResults]
+        datasetResult.append({
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:24]: I would not add this data to a list first, but rather call `writer.writerow` directly for this data

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        writer.writeheader()
+        writer.writerows(resultsForCsv)
+
+    if taskRun.createArtifact(csvSamplesPath, csvSamplesPath.name) is None:
+        logging.warning(">> [Image Segmentation] Failed to upload csv file with samples results as artifact")
+    else:
+        logging.info(">> [Image Segmentation] The csv file with samples results has been uploaded as artifact")
+
+
+def createCsvResults(taskRun: TaskRun[ImageDataset], sampleResults: list[dict[str, str]]) -> None:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:67]: Is there a need for this to be a separate function?

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+
+def uploadSampleArtifacts(taskRun: TaskRun[ImageDataset], sampleId: int, plotPath: Path, sampleResult: dict[str, str]) -> None:
+    csvSamplePath = folder_manager.temp.joinpath(f"{sampleId}.csv")
+    with csvSamplePath.open("w") as file:
+        writer = csv.DictWriter(file, fieldnames = sampleResult.keys())
+        writer.writeheader()
+        writer.writerow(sampleResult)
+
+    if taskRun.createArtifact(csvSamplePath, f"sample_results/{sampleId}/{sampleId}.csv") is None:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:82]: Let's call the file `sample_results/{sampleId}/results.csv`

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    for polygon in polygons:
+        draw = ImageDraw.Draw(image)
+        draw.polygon(polygon, fill = 1)
+
+    return np.array(image)
+
+
+def processResult(
+    taskRun: TaskRun[ImageDataset],
+    result: Results
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:142]: This function must have sample as a parameter

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            polygon = mask.xyn
+            for points in polygon:
+                poly = points.tolist()
+                poly = [tuple(point) for point in poly]
+                poly = [(int(x * width), int(y * height)) for (x, y) in poly]
+                polygons.append(poly)
+
+        classIds = [int(x) for x in result.boxes.cls.tolist()]
+        classNames: dict[int, str] = result.names
+        setOfClassIds = list(set(classIds))
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:176]: There is no need for this to be a variable

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    csvRowResult["sample id"] = f"{sampleId}"
+    csvRowResult["sample name"] = sample.name
+    for className in taskRun.dataset.classes.labels:
+        csvRowResult[className] = "0.00"
+
+    polygons: list[list[tuple[int, int]]] = []
+    height = result.orig_shape[0]
+    width = result.orig_shape[1]
+    if result.masks is not None:
+        for mask in result.masks:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:166]: I'd extract this into a separate function

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+                poly = [tuple(point) for point in poly]
+                poly = [(int(x * width), int(y * height)) for (x, y) in poly]
+                polygons.append(poly)
+
+        classIds = [int(x) for x in result.boxes.cls.tolist()]
+        classNames: dict[int, str] = result.names
+        setOfClassIds = list(set(classIds))
+
+        for classId in setOfClassIds:
+            indexes = [i for i, clsId in enumerate(classIds) if clsId == classId]
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:179]: calculating accuracy of a single class should be a separate function

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+    logging.info(f">> [Image Segmentation] The {str(device).upper()} will be used for validating")
+
+    logging.info(">> [Image Segmentation] Validatin the model")
+    results = model.predict(
+        source = samples,
+        save = False,
+        batch = taskRun.parameters["batchSize"],
+        imgsz = imgSize,
+        plots = False,
+        conf = 0.1
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:215]: Should we make this a parameter of the Task?

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: tasks/image-segmentation-yolov8/main.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    if taskRun.parameters["validation"]:
+        logging.info(">> [Image Segmentation] Validating mode")
+        if taskRun.parameters["trainedModel"] is None:
+            raise RuntimeError("Model used for image segmentation that needs validation is not valid")
+
+        ctxModel: Model = taskRun.parameters["trainedModel"]
+        ctxModel.download()
+
+        with (ctxModel.path / ctxModel.modelDescriptorFileName()).open("r") as file:
+            modelDesc = dict(json.load(file))
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/main.py:25]: `json.load` will return a `list` or `dict`. You should invoke the `dict` constructor with the return value, but rather check the type of the returned object using `isinstace` function.

File: tasks/image-segmentation-yolov8/main.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            imgSize = int(modelDesc["imageSize"])
+            classLabels = [str(clazz["label"]) for clazz in modelDesc["labels"]]
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/main.py:27]: Same here for `clazz["label"]`

File: tasks/image-segmentation-yolov8/main.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+    if taskRun.parameters["validation"]:
+        logging.info(">> [Image Segmentation] Validating mode")
+        if taskRun.parameters["trainedModel"] is None:
+            raise RuntimeError("Model used for image segmentation that needs validation is not valid")
+
+        ctxModel: Model = taskRun.parameters["trainedModel"]
+        ctxModel.download()
+
+        with (ctxModel.path / ctxModel.modelDescriptorFileName()).open("r") as file:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/main.py:24]: Do not use parenthessis to wrap a result of `/` operator on `pathlib.Path`. Use `pathlib.Path.joinpath` function instead.

File: tasks/image-segmentation-yolov8/main.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+            modelDesc = dict(json.load(file))
+            imgSize = int(modelDesc["imageSize"])
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/main.py:26]: Same here

File: tasks/image-segmentation-yolov8/main.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+        taskRun.updateStatus(TaskRunStatus.inProgress, "Downloading dataset")
+        taskRun.dataset.download()
+
+        validate(taskRun, modelPath, imgSize)
+    else:
+        logging.info(">> [Image Segmentation] Training mode")
+        if taskRun.parameters["epochs"] is None:
+            raise RuntimeError(">> [Image Segmentation] The number of training epochs is not defined")
+        if taskRun.parameters["imageSize"] is None:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/main.py:43]: Use `get` instead of `[]` if parameter is optional (not required) in the `task.yaml` file

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+                "Class": label,
+                "Accuracy": f"{np.mean(classAcc):.2f}",
+                "Accuracy stdev": f"{np.std(classAcc):.2f}"
+            })
+
+        writer.writerow({
+        "Class": "dataset",
+        "Accuracy": f"{np.mean(datasetAcc):.2f}",
+        "Accuracy stdev": f"{np.std(datasetAcc):.2f}"
+    })
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:42]: Formatting is incorrect

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+    if sampleAnnotation is None:
+        raise RuntimeError(f">> [Image Segmentation] The sample named \"{sample.name}\" (id: \"{sample.id}\") has no annotation.")
+
+    csvRowResult: dict[str, str] = {}
+    csvRowResult["sample id"] = f"{sample.id}"
+    csvRowResult["sample name"] = sample.name
+    for className in taskRun.dataset.classes.labels:
+        csvRowResult[className] = "0.00"
+
+
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:179]: 2 newlines

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+def processSampleResult(
+    taskRun: TaskRun[ImageDataset],
+    result: Results,
+    sample: ImageSample
+) -> dict[str, str]:
+
+    sampleAnnotation = sample.load().annotation
+    if sampleAnnotation is None:
+        raise RuntimeError(f">> [Image Segmentation] The sample named \"{sample.name}\" (id: \"{sample.id}\") has no annotation.")
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:171]: If a sample has no annotation that means its groundtruth mask is zeroes (background)

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+
+    taskRun.updateStatus(TaskRunStatus.inProgress, "Processing predictions")
+    processedResults: list[dict[str, str]] = []
+    with ProcessPoolExecutor(max_workers = mp.cpu_count()) as executor:
+        futures: list[Future[dict[str, str]]] = []
+        for result, sample in zip(results, taskRun.dataset.samples):
+            future = executor.submit(processSampleResult, taskRun, result, sample)
+            futures.append(future)
+
+
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:237]: 2 newlines

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+        save = False,
+        batch = taskRun.parameters["batchSize"],
+        imgsz = imgSize,
+        plots = False,
+        conf = taskRun.parameters["confidenceTreshold"]
+    )
+
+    taskRun.updateStatus(TaskRunStatus.inProgress, "Processing predictions")
+    processedResults: list[dict[str, str]] = []
+    with ProcessPoolExecutor(max_workers = mp.cpu_count()) as executor:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:231]: add newline above the with statement

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+            futures.append(future)
+
+
+        for counter, future in enumerate(as_completed(futures)):
+            try:
+                processedResults.append(future.result())
+            except FileNotFoundError as e:
+                logging.warning(e)
+
+            logging.info(f">> [Image Segmentation] Processing results for sample {counter + 1}/{taskRun.dataset.count} has been finished")
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:244]: Can you print the sample accuracy as well? And dataset accuracy at the end.

**Review by dule1322 — CHANGES_REQUESTED** [review]
